src/core/main.c

```diff
@@ -1253,37 +1253,39 @@ static void bump_file_max_and_nr_open(void) {
 #endif
 
 #if BUMP_PROC_SYS_FS_NR_OPEN
-        int v = INT_MAX;
+        unsigned int v = INT_MAX;
 
-        /* Argh! The kernel enforces maximum and minimum values on the fs.nr_open, but we don't really know
-         * what they are. The expression by which the maximum is determined is dependent on the architecture,
-         * and is something we don't really want to copy to userspace, as it is dependent on implementation
-         * details of the kernel. Since the kernel doesn't expose the maximum value to us, we can only try
-         * and hope. Hence, let's start with INT_MAX, and then keep halving the value until we find one that
-         * works. Ugly? Yes, absolutely, but kernel APIs are kernel APIs, so what do can we do... 🤯 */
+        // cf. https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/fs/file.c?h=v6.8#n27
+	#define BITS_PER_LONG __WORDSIZE
+        #define __const_min(x, y) ((x) < (y) ? (x) : (y))
+        unsigned int sysctl_nr_open_max =
+            __const_min(INT_MAX, ~(size_t)0/sizeof(void *)) & -BITS_PER_LONG;
```

Comment:
can someone explain to me what `& ~BITS_PER_LONG does, I am not sure I grok this.


please use SIZE_MAX instead of `~(size_t)0`

Reply:
>can someone explain to me what `& ~BITS_PER_LONG does, I am not sure I grok this.

It zeroes out the last ~`log2(BITS_PER_LOG)` bits from the value of `INT_MAX`. The effect is to ensure that there aren't any overflows for `sysctl_nr_open_max`. I really don't want to change this expression relative to the kernel source, especially since it hasn't changed in almost 16 years.

Cf. https://github.com/torvalds/linux/commit/eceea0b3df05ed262ae32e0c6340cc7a3626632d

Reply:
> please use SIZE_MAX instead of ~(size_t)0

I would prefer to avoid diverging from kernel source. This expression exists as-is in `fs/file.c` and I'm extremely opposed to changing it. Please also see https://github.com/systemd/systemd/pull/32740#discussion_r1596403025 .